gtsam/nonlinear/FunctorizedFactor.h

```diff
@@ -0,0 +1,145 @@
+/* ----------------------------------------------------------------------------
+
+ * GTSAM Copyright 2010, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * Authors: Frank Dellaert, et al. (see THANKS for the full author list)
+
+ * See LICENSE for the license information
+
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file FunctorizedFactor.h
+ * @date May 31, 2020
+ * @author Varun Agrawal
+ **/
+
+#pragma once
+
+#include <gtsam/base/Testable.h>
+#include <gtsam/nonlinear/NonlinearFactor.h>
+
+#include <cmath>
+
+namespace gtsam {
+
+/**
+ * Factor which evaluates functor and uses the result to compute
+ * error on provided measurement.
+ * The provided FUNCTOR should provide two type aliases: `argument_type` which
+ * corresponds to the type of input it accepts and `return_type` which indicates
+ * the type of the return value. This factor uses those type values to construct
+ * the functor.
+ *
+ * Template parameters are
+ * @param FUNCTOR: A class which operates as a functor.
+ * 
+ * Example:
+ *   Key key = Symbol('X', 0);
+ *   auto model = noiseModel::Isotropic::Sigma(9, 1);
+ *
+ *   /// Functor that takes a matrix and multiplies every element by m
+ *   class MultiplyFunctor {
+ *     double m_; ///< simple multiplier
+ *    public:
+ *     using argument_type = Matrix;
+ *     using return_type = Matrix;
+ *     MultiplyFunctor(double m) : m_(m) {}
+ *     Matrix operator()(const Matrix &X,
+ *              OptionalJacobian<-1, -1> H = boost::none) const {
+ *       if (H) *H = m_ * Matrix::Identity(X.rows()*X.cols(), X.rows()*X.cols());
+ *       return m_ * X;
+ *     }
+ *   };
+ *
+ *   Matrix measurement = Matrix::Identity(3, 3);
+ *   double multiplier = 2.0;
+ *   FunctorizedFactor<MultiplyFunctor> factor(keyX, measurement, model, multiplier);
+ */
+template <typename FUNCTOR>
```

Comment:
To address my other comment, perhaps we should instead template on T. And then a Functor/lambda/std::function passed needs to take (T, jacobian). Not necessarily dynamic jacobian, btw...

Reply:
The problem here is that we won't know what the return type is since we don't have anything to call `decltype` on. We need the return type to define the measurement type.